sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2963,7 +2963,11 @@ class Analyzer(override val catalogManager: CatalogManager)
 
     private def getNondeterToAttr(exprs: Seq[Expression]): Map[Expression, NamedExpression] = {
       exprs.filterNot(_.deterministic).flatMap { expr =>
-        val leafNondeterministic = expr.collect { case n: Nondeterministic => n }
+        val leafNondeterministic = expr.collect {
+          case n: Nondeterministic => n
+          case e: CallMethodViaReflection => e
+          case udf: UserDefinedExpression if !udf.deterministic => udf
```

Comment:
shall we just write `case e if !e.deterministic => e`?

Reply:
First time, I tried this [33eda3c](https://github.com/apache/spark/pull/30703/commits/33eda3cb2575f1d11f013b87cd4196b7668db8ce), but the behavior is unexpected.

Let's say we have a sql `select * from t order by rand()`, then it will pull out `SortOrder(rand())` into `Project`.

Reply:
ah, so we need to collect leaf non-deterministic nodes here. Shall we create a trait for them?

Reply:
I'd like to add a `LeafNondeterministic`, but seems there is not only leaf node, like `Rand(child)`.

Reply:
Can we make `CallMethodViaReflection` to extend `Nondeterministic`? I think `UserDefinedExpression` is the only exception that its deterministic is a flag.

Reply:
ok. try it.